lib/ohai/plugins/openstack.rb

```diff
@@ -18,53 +19,67 @@
 require "ohai/mixin/ec2_metadata"
 
 Ohai.plugin(:Openstack) do
+  include Ohai::Mixin::Ec2Metadata
+
   provides "openstack"
+  depends "dmi"
+  depends "etc"
 
-  include Ohai::Mixin::Ec2Metadata
+  # do we have the openstack dmi data
+  def openstack_dmi?
+    # detect a manufacturer of OpenStack Foundation
+    if dmi[:system][:all_records][0][:Manufacturer] =~ /OpenStack/
+      Ohai::Log.debug("Plugin Openstack: has_openstack_dmi? == true")
+      true
+    end
+  rescue NoMethodError
+    Ohai::Log.debug("Plugin Openstack: has_openstack_dmi? == false")
+    false
+  end
+
+  # check for the ohai hint and log debug messaging
+  def openstack_hint?
+    if hint?("openstack")
+      Ohai::Log.debug("Plugin Openstack: openstack hint present")
+      return true
+    else
+      Ohai::Log.debug("Plugin Openstack: openstack hint not present")
+      return false
+    end
+  end
 
-  def collect_openstack_metadata(addr = Ohai::Mixin::Ec2Metadata::EC2_METADATA_ADDR, api_version = "2013-04-04")
-    path = "/openstack/#{api_version}/meta_data.json"
-    uri = "http://#{addr}#{path}"
+  # dreamhost systems hae the dhc-user on them
+  def openstack_provider
     begin
-      response = http_client.get_response(URI.parse(uri), nil, nil)
-      case response.code
-      when "200"
-        FFI_Yajl::Parser.parse(response.body)
-      when "404"
-        Ohai::Log.debug("Encountered 404 response retreiving OpenStack specific metadata path: #{path} ; continuing.")
-        nil
-      else
-        raise "Encountered error retrieving OpenStack specific metadata (#{path} returned #{response.code} response)"
+      return "dreamhost" if etc["passwd"]["dhc-user"]
+    rescue NoMethodError
+      # handle etc not existing on non-linux systems
+    end
+    return "openstack"
+  end
```

Comment:
I had it that way and it ends up returning nil when it can't find the key, which isn't what you want.

Scenarios:
etc plugin doesn't exist at all and the rescue is necessary -> openstack
etc plugin exists, but dhc-user doesn't exist -> openstack
etc plugin exists and dhc-user does exist -> dreamhost